packages/contracts/contracts/protocol/Exchange/MixinExchangeCore.sol

```diff
@@ -354,7 +354,7 @@ contract MixinExchangeCore is
                     order.makerAddress,
                     signature
                 ),
-                "INVALID_ORDER_SIGNATURE"
+                string(abi.encodeWithSignature("INVALID_ORDER_SIGNATURE(bytes32)", orderInfo.orderHash))
```

Comment:
This is a good hack to add data to the existing revert reasons. Remember that Solidity encodes those as `abi.encodeWithSignature("Error(string)", <revert reason>)`. So now we effectively get

```
abi.encodeWithSignature("Error(string)", string(
    abi.encodeWithSignature("INVALID_ORDER_SIGNATURE(bytes32)",
        orderInfo.orderHash)))
```

This can lead to zero bytes causing C-style string termination. Please test to make sure all this is handled correctly by all libraries involved.

I expect some future version of Solidity to support something natively. There is currently no open issue for it, so we could propose one.

Whatever the future standard is going to be, it is unlikely to be our current double-encoded one. Rather it would be something like:

```
abi.encodeWithSignature("ErrorInvalidOrderSignature(bytes32)", orderInfo.orderHash)
```

We could implement this already by writing it in assembly and it would be cleaner and cheaper, but that doesn't seem worth (1) the effort and (2) breaking compatibility with the existing standard.

Speaking of compatibility, the current implementation does not produce a readable string, which is what is expected. I do not think this is a major concern.

Reply:
Very good catch. I think the initial EIP about revert reasons had some discussions about additional data but they've decided to not do that for now because of 1. It's easier. 2. It's faster to decide on smth easier 3. Not everyone needs additional data (debatable).
I think in future they might agree on smth but there is no way for us to guess. Assuming they will not break the basic functionality - there will always be a way for our users to decode our error messages. If you can propose a better scheme - I'm all about it. I think I'm against the assembly solution because it's harder for users to see what's happening. We might still write it and check the gas  costs though. It might break the existing libraries trying to decode the standard revert reasons though. Do I understand it correctly, that we'll need to write a separate assembly code for each revert reason type? They have different payloads.

Reply:
> It might break the existing libraries trying to decode the standard revert reasons though.

Yeah. That's why we should stick with what you made for now. We could put up an EIP and push some standard through, then switch in v.3.

> We might still write it and check the gas costs though.

Generally you would not optimize the error cases. Do we need cheap reverts for batchfills?

> Do I understand it correctly, that we'll need to write a separate assembly code for each revert reason type?

Just one assembly function would be required:

```solidity
function rawRevert(bytes data) internal {
     assembly {
        revert(add(data, 0x20), mload(data))
     }
}
```

which can then be used as:

```
rawRevert(abi.encodeWithSignature("ErrorInvalidOrderSignature(bytes32)", orderInfo.orderHash))
```

Reply:
> This can lead to zero bytes causing C-style string termination. Please test to make sure all this is handled correctly by all libraries involved.

Another thing to watch out for are byte sequences that are not valid UTF8. Casting binary to strings is a dangerous.